mne/preprocessing/faster.py

```diff
@@ -0,0 +1,463 @@
+
+
+def _power_gradient(data, sfreq, prange):
+    """Estimate the gradient of the power spectrum at upper frequencies.
+
+    Parameters
+    ----------
+    data : N-D numpy array
+        The timeseries to estimate signal power for. The last dimension
+        is presumed to be time.
+    sfreq : float
+        The sample rate of the timeseries.
+    prange : pair of floats
+        The (lower, upper) frequency limits of the power spectrum to use. In
+        the FASTER paper, they set these to the passband of the lowpass filter.
+
+    Returns
+    -------
+    grad : N-D numpy array
+        The gradients of each timeseries.
+    """
+    fs, ps = _efficient_welch(data, sfreq)
+
+    # Limit power spectrum to upper frequencies
+    start, stop = (np.searchsorted(fs, p) for p in prange)
+    if start >= ps.shape[1]:
+        raise ValueError(("Sample rate insufficient to estimate {} Hz power. "
+                          "Use the 'power_gradient_range' parameter to tweak "
+                          "the tested frequencies for this metric or use the "
+                          "'metrics' parameter to disable the "
+                          "'power_gradient' metric.").format(prange[0]))
+    ps = ps[:, start:stop]
+
+    # Compute mean gradients
+    return np.mean(np.diff(ps), axis=1)
+
+
+def _deviation(data):
+    """Computes the deviation from mean for each channel in a set of epochs.
+
+    This is not implemented as a lambda function, because the channel means
+    should be cached during the computation.
+
+    Parameters
+    ----------
+    data : 3D numpy array
+        The epochs (#epochs x #channels x #samples).
+
+    Returns
+    -------
+    dev : 1D numpy array
+        For each epoch, the mean deviation of the channels.
+    """
+    ch_mean = np.mean(data, axis=2)
+    return ch_mean - np.mean(ch_mean, axis=0)
+
+
+def faster_bad_channels(epochs, picks=None, thres=3, use_metrics=None,
+                        max_iter=2, return_by_metric=False):
+    """Implements the first step of the FASTER algorithm.
+
+    This function attempts to automatically mark bad EEG channels by performing
+    outlier detection. It operated on epoched data, to make sure only relevant
+    data is analyzed.
+
+    Parameters
+    ----------
+    epochs : Instance of Epochs
+        The epochs for which bad channels need to be marked
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
+    thres : float
+        The threshold value, in standard deviations, to apply. A channel
+        crossing this threshold value is marked as bad. Defaults to 3.
+    use_metrics : list of str
+        List of metrics to use. Can be any combination of:
+            'variance', 'correlation', 'hurst', 'kurtosis', 'line_noise'
+        Defaults to all of them.
+    max_iter : int
+        The maximum number of iterations performed during outlier detection
+        (defaults to 2, original FASTER paper uses 1).
```

Comment:
I'd like to avoid commitments to internal outlier functions. It's been a mistake by the way to make bads.find_outliers public. We should first of all use the original plain vanilla FASTER to evaluate it.